src/App.tsx

```diff
@@ -1,24 +1,311 @@
-import { UserWarning } from './UserWarning';
+import React, {
+  useEffect,
+  useState,
+  ChangeEvent,
+  FormEvent,
+} from 'react';
 
-const USER_ID = 0;
+import { Header } from './components/Header';
+import { Main } from './components/Main';
+import { Footer } from './components/Footer';
+import { ErrorPanel } from './components/ErrorPanel';
+
+import { Todo } from './types/Todo';
+import { Filters } from './types/Filters';
+import { Errors } from './types/Errors';
+
+import {
+  getTodos,
+  addTodo,
+  removeTodo,
+  updateTodoStatus,
+  updateTodoTitle,
+} from './api/todos';
+
+const USER_ID = 6160;
 
 export const App: React.FC = () => {
-  if (!USER_ID) {
-    return <UserWarning />;
-  }
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [error, setError] = useState<Errors>(Errors.NoError);
+  const [filteredTodos, setFilteredTodos] = useState<Todo[]>(todos);
+  const [selectedFilter, setSelectedFilter] = useState<Filters>(Filters.All);
+  const [newTodoTitle, setNewTodoTitle] = useState('');
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [loadingTodoIds, setLoadingTodoIds] = useState<number[]>([]);
+  const [isNewTodoLoading, setIsNewTodoLoading] = useState(false);
+
+  const activeTodos = todos.filter(todo => !todo.completed);
+  const completedTodos = todos.filter(todo => todo.completed);
+
+  const saveEditedTitle = (
+    todoId: number,
+    todoTitle: string,
+    setEditingTodoId: (value: number) => void,
+  ) => {
+    setLoadingTodoIds([...loadingTodoIds, todoId]);
+    updateTodoTitle(todoId, todoTitle)
+      .then(() => {
+        getTodos(USER_ID)
+          .then(todosFromServer => {
+            setTodos(todosFromServer);
+            setFilteredTodos(todosFromServer);
+            setLoadingTodoIds([]);
+            setEditingTodoId(0);
+          })
+          .catch(() => {
+            setError(Errors.CantUpdate);
+          });
+      });
+  };
+
+  const toggleTodoStatus = (todoId: number, completed: boolean) => {
+    updateTodoStatus(todoId, !completed)
+      .then(() => {
+        setLoadingTodoIds([...loadingTodoIds, todoId]);
+        getTodos(USER_ID)
+          .then(todosFromServer => {
+            setTodos(todosFromServer);
+            setFilteredTodos(todosFromServer);
+            setLoadingTodoIds([]);
+          })
+          .catch(() => {
+            setError(Errors.CantUpdate);
+          });
+      });
+  };
+
+  const getUpdatedTodos = (
+    todosToUpdate: Todo[],
+    everyTodoCompleted: boolean,
+    noTodoCompleted: boolean,
+    someTodosCompleted: boolean,
+  ) => {
+    const updatedTodos: Todo[] = [];
+
+    todosToUpdate.forEach(todo => {
+      let updatedTodo = { ...todo };
+
+      if (someTodosCompleted && !todo.completed) {
+        updatedTodo = { ...todo, completed: !todo.completed };
+      }
+
+      if (everyTodoCompleted || noTodoCompleted) {
+        updatedTodo = { ...todo, completed: !todo.completed };
+      }
```

Comment:
could be shorter - update props when constructing updatedTodo
```suggestion
      const completed = (someTodosCompleted && !todo.completed)
        || (everyTodoCompleted || noTodoCompleted)
      const updatedTodo = {
        ...todo,
        completed,
      };
```

Reply:
thank you for your comments, i created a nice and common handlerFunction for both 'blur' and ' form submit' but with this one i had hard times shortening it.
the above described case updates all of them even when there are some completed and some active todos so i think i cant go for it. 
or i am just missing something

Reply:
oh, so you mean that `everyTodoCompleted || noTodoCompleted` are dependent from `someTodosCompleted && !todo.completed`?
if so, you could carry on.

I think you should just try it and test out edge-cases that should be handled by the logic.

Have you tried to apply the suggestion? And test it on the rendered page?

Reply:
yes of course but then i realized that i probably need either switch or if statements because sometimes on only want to update some of them (both on the page and in the api)